Support import of raw cca tokens with C_ObjectCreate

hfreude opened the pull request:
3 commits:
- TESTCASES: new testcase to test CCA opaque key token export and import
- CCA: support import of raw cca tokens with C_ObjectCreate via CKA_IBM_OPAQUE
- CCA: store only public token in IBM_OPAQUE attribute
for more details pleas see header of each commit

hfreude commented:
PR is updated, now without AES cipher key support and without requirement for CKA_VALUE for des and aes keys.

ifranzki commented on usr/lib/common/new_host.c:
```diff
@@ -3638,7 +3638,7 @@ CK_RV SC_GenerateKeyPair(STDLL_TokData_t *tokdata,
 
     TRACE_DEBUG("Private Template:\n");
-    attr = pPublicKeyTemplate;
+    attr = pPrivateKeyTemplate;
```
Please check if this has to be corrected in the other copies of new_host.c as well (ep11 token and icsf token).

hfreude replied:
needs also to get fixed in the other new_host.c files ...
I'll do that - maybe split these fixes into a new commit.

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
 
+#ifdef DEBUG
+/* Debug function: hexdump buffer with TRACE_DEBUG */
+static void dump_mem(const char *prestr, void *buf, unsigned len)
```
In ep11_specific.c there is a macro defined to dump memory:
`#define TRACE_DEBUG_DUMP(_buf, _buflen) hexdump(_buf, _buflen)`
Function hexdump() is alos local there. 
Maybe we can add a function into trace.c and a macro like TRACE_DEBUG_DUMP to use it, so that it can be used everywhere?

hfreude replied:
yes, makes sense - will do this

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+}
+
+/* Debug function: dump one attribute via TRACE_DEBUG */
+static void dump_attr(CK_ATTRIBUTE *a)
```
This should probably go into attributes.c or p11util.c so that it is available for others, too. 
In p11utilc, there are already functions p11_get_ckr() and p11_get_ckm() to get the textual representation of return codes and mechanisms. maybe it would make sense to add a p11_get_cka() as well?

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+}
+
+/* Debug function: dump list of attribues from a template via TRACE_DEBUG */
+static void dump_template(TEMPLATE *tmpl)
```
move to template.c, or p11util.c ?

hfreude replied:
done, will be part of the next PR update

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+                                      enum cca_token_type *keytype,
+                                      unsigned int *keybitsize)
+{
+    if (t[0] == 0x01 && (t[4] == 0x00 || t[4] == 0x01)) {
```
I usually like struct to overlay over the buffer better than hard coded offsets, but since the CCA token uses the offset-way already, I am fine with this, too.

hfreude replied:
I have the structs - but as most of the cca specific code deals with some kind of offsets and would need a total rework to adapt to structs, I've chosen to use these simple offset values.

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -239,6 +240,300 @@ static const CK_ULONG cca_mech_list_len =
+        return TRUE;
+    }
+
+    if (t[0] == 0x01 && t[4] == 0x04) {
```
2 levels of switch()-es would be easier to read maybe:
```
switch (t[0]) {
case 0x01:
    switch (t[4]) {
    case xxx:
    }
}
```

hfreude replied:
well, no
the idea of the whole function is to group it based on keytypes. Mixing of ifs and switches and more levels may give the function a 1us performance boost but the human readability is much more worse. At least this is my understanding.

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -3988,331 +4607,553 @@ static CK_RV build_public_EC_key_value_structure(CK_BYTE *pubkey, CK_ULONG puble
+        return CKR_FUNCTION_FAILED;
+    }
 
+    /* we only deal with key objects here */
```
Actually, you could check here already if an CKA_IBM_OPAQUE is there, and if so, then analyze it and find out about the key type and class from the token. This would lift the requirement for the user to specify CLA_KEY_TYPE and CKA_CLASS when importing an opaque key. The key token does in fact contain all this information already, so no need for the user to specify it.
The only question would be if an EP11 key token also allows to get the same kind of information from it. So once we do the key blob import for EP11, it then should behave the same as for CCA.

hfreude replied:
I'll leave this to Reinhard to decide. It's doable but then a user does not need and provide any key information but only the CCA secure key token. So no way to check if this was the key type the user intended to import.

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -4409,6 +5250,11 @@ CK_RV token_specific_generic_secret_key_gen(STDLL_TokData_t * tokdata,
 
+#ifdef DEBUG
+    TRACE_DEBUG("%s: imported object template attributes:\n", __func__);
+    dump_template(template);
```
If you add this here (which makes sense, now that we have this functions), why not also adding it to the other keygen functions as well?
Any maybe adjust the trace message, here is is not an 'imported' key, but a 'generated' one.

ifranzki commented on testcases/misc_tests/cca_export_import_test.c:
```diff
@@ -0,0 +1,1440 @@
+/*
```
I guess we should have the standard file header here with copyright infos, license, etc.

hfreude replied:
Ok, will do.

hfreude commented:
Ok, hopefully this was the last update of the PR:
- added some more attributes for the dump_attr function.
- added CKA_SENSITIVE=true for the import of sym and priv keys (request from Reinhard)

ifranzki commented on usr/lib/cca_stdll/cca_specific.c:
```diff
@@ -238,6 +238,27 @@ static const MECH_LIST_ELEMENT cca_mech_list[] = {
+/* Helper function: build attribute and update template */
+static CK_RV build_update_attribute(TEMPLATE * tmpl,
+                                    CK_ATTRIBUTE_TYPE type,
+                                    CK_BYTE * data, CK_ULONG data_len)
```
We could move this function to common code (e.g. utility.c, were build_attribute() is already) so that others could also utilize it, since many other places also have the sequence of build_attribute() followed by template_update_attribute(), but that's not something that should hold off this PR.

hfreude replied:
yes, this could and should be done. And - as this function avoids the free() if update_attribute() fails, all the places where build_attribute() + update_attribute() is used would benefit from this.

ifranzki commented on testcases/misc_tests/cca_export_import_test.c:
```diff
@@ -0,0 +1,1448 @@
+#include "common.c"
+
+// define this to enable the AES cipher key import test:
+// #define CCA_AES_CIPHER_KEY_SUPPORTED
```
I don't find this define anywhere else except in this testcase, especially not in cca_specific.c. Is this a left over, or did you leave it there intentionally?

hfreude replied:
You are right. This is local to the file.
The code to support import of CCA AES cipher blobs is easy - just add it to the switch within the symetric key import code function. However, I did not want to discard the test code as I spend some time to develop it and so I commented it out with this #ifdef CCA_AES_CIPHER_KEY_SUPPORTED

ifranzki reviewed (approved):
lgtm

p-steuer reviewed (approved):
lgtm
